packages/babel-parser/src/plugins/typescript/index.ts

```diff
@@ -3165,21 +3125,93 @@ export default (superClass: ClassWithMixin<typeof Parser, IJSXParserMixin>) =>
       flags: ParseStatementFlag,
       decorators?: N.Decorator[] | null,
     ): N.Statement {
-      if (this.match(tt._const) && this.isLookaheadContextual("enum")) {
-        const node = this.startNode<N.TsEnumDeclaration>();
-        this.expect(tt._const); // eat 'const'
-        return this.tsParseEnumDeclaration(node, { const: true });
-      }
-
-      if (this.isContextual(tt._enum)) {
-        return this.tsParseEnumDeclaration(
-          this.startNode<N.TsEnumDeclaration>(),
-        );
-      }
-
-      if (this.isContextual(tt._interface)) {
-        const result = this.tsParseInterfaceDeclaration(this.startNode());
-        if (result) return result;
+      if (!this.state.containsEsc) {
+        switch (this.state.type) {
+          case tt._const: {
+            if (this.isLookaheadContextual("enum")) {
+              const node = this.startNode<N.TsEnumDeclaration>();
+              this.expect(tt._const); // eat 'const'
+              return this.tsParseEnumDeclaration(node, { const: true });
+            }
+            break;
+          }
+          case tt._abstract:
+          case tt._declare: {
+            if (
+              this.nextTokenIsIdentifierAndNotTSRelationalOperatorOnSameLine()
```

Comment:
There is a finite number of keyword tokens allowed to follow `abstract` / `declare`. In reality most input source is valid, to simplify the lookahead logic and exit early for valid code, we check all identifiers except `in`, `instanceof`, `as` and `satisifies`: If a declaration can not be parsed, it must be an invalid production in _both_ JS and TS.